tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreManagedLedgerOffloader.java

```diff
@@ -621,4 +634,75 @@ public void close() {
             }
         }
     }
+
+    @Override
+    public void scanLedgers(OffloadedLedgerMetadataConsumer consumer, Map<String, String> offloadDriverMetadata) throws ManagedLedgerException {
+        BlobStoreLocation bsKey = getBlobStoreLocation(offloadDriverMetadata);
+        String endpoint = bsKey.getEndpoint();
+        String readBucket = bsKey.getBucket();
+        log.info("Scanning bucket {}, bsKey {}, location {} endpoint{} ", readBucket, bsKey,
+                config.getBlobStoreLocation(), endpoint);
+        BlobStore readBlobstore = blobStores.get(config.getBlobStoreLocation());
+        int batchSize = 100;
+        String bucketName = config.getBucket();
+        String marker = null;
+        do {
+            marker = scanContainer(consumer, readBlobstore, bucketName, marker, batchSize);
+        } while (marker != null);
+
+    }
+
+    private String scanContainer(OffloadedLedgerMetadataConsumer consumer, BlobStore readBlobstore,
+                                 String bucketName,
+                                 String lastMarker,
+                                 int batchSize) throws ManagedLedgerException {
+        ListContainerOptions options = new ListContainerOptions()
+                .maxResults(batchSize)
+                .withDetails();
+        if (lastMarker != null) {
+            options.afterMarker(lastMarker);
+        }
+        PageSet<? extends StorageMetadata> pages = readBlobstore.list(bucketName, options);
+        for (StorageMetadata md : pages) {
+            log.info("Found {} ",md);
+            String name = md.getName();
+            Long size = md.getSize();
+            Date lastModified = md.getLastModified();
+            StorageType type = md.getType();
+            if (type != StorageType.BLOB) {
+                continue;
+            }
+            URI uri = md.getUri();
+            Map<String, String> userMetadata = md.getUserMetadata();
```

Comment:
Looks like you are trying to get the ledger metadata from the Blob metadata? IIRC, the ledger metadata was saved in the index file. You will get nothing from here

Reply:
Yes, the ledger metadata stored in index file. https://github.com/apache/pulsar/blob/02eb31b372b2bf72350e8f6cbab552e1627e6197/tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreManagedLedgerOffloader.java#L172-L174

`md.getUserMetadata()` can only get the following properties.
https://github.com/apache/pulsar/blob/02eb31b372b2bf72350e8f6cbab552e1627e6197/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java#L1259-L1262

Reply:
in this patch I added that information
https://github.com/apache/pulsar/pull/14907
see here
https://github.com/apache/pulsar/pull/14907/files#diff-10c82eb07e9dae9051b656297cf7dc38d051aad4b61b8555d076d56b4030b5bdR188

Reply:
The ledger info doesn't stored in `extraMetadata`, but in `readHandle.getLedgerMetadata()`

Reply:
here
https://github.com/apache/pulsar/blob/2f7890442f5a75695330ab5f6f4331afe65a5d8f/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2954

I will show you some examples of the results